### DISCUSSION TITLE
How to implement multiple approvals (reviews) for syncing in ArgoCD

### DISCUSSION THREAD

**Discussion by sl3dg3hamm3r** [argoproj/argo-cd#17540] [Q&A] [answered]
Disclaimer: I asked the question on SO (https://stackoverflow.com/questions/78166296/how-to-implement-multiple-approvals-for-syncing-in-argocd), but I guess it is the better place to ask it here ...

I thought that's an easy one, but could not find any resource yet describing it. On a production stage, there should be more than one person validating the out-of-sync diffs and hitting (reviewing and approving) the button for syncing for any registered application in ArgoCD. So, the resource should be only synced after two separate persons approved it.

How could this be implemented?

**Comment by vorozhko** [comment, marked as answer]
ArgoCD is not  suited for that. It much better to design approval locgic in control version systesm like Github or Gitlab.

**Reply by torfjor** [reply]
I'd like to disagree. 

If you take out the automatic sync, you have a golden opportunity to make a manual approval step with as many required 👍🏻 as you could possibly need. 

The notifications controller already gives you the opportunity of sending a webhook request to a third-party component whenever an application that matches the criteria goes OutOfSync. 

That third-party component can be fed the application's status to figure out the observed and desired revision. The status even has typed hints for which resources have changed, which gives you an excellent opportunity to make automatic approvals for changes you deem "safe". 

Once that third-party component has figured out it's time to sync, use the the Argo CD API to sync the app, or express the intent through the [Kubernetes API](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-kubectl/).

**Reply by crenshaw-dev** [reply]
You can do all that. But you're rebuilding stuff that has existed in SCMs for years.

**Reply by torfjor** [reply]
Yeah, I don't trust my approvers to know the full extent of what a change in a base or component touches in an n-level deep overlay. If you already use a flow with pre-rendered manifests I guess the point is moot, but then you'd still need to have a lot of CI machinery to make that process smooth. When are we getting manifest hydration again? 😅

**Reply by crenshaw-dev** [reply]
Yep, makes sense. Source hydrator is out: https://argo-cd.readthedocs.io/en/stable/user-guide/source-hydrator/

It has the ability to `hydrateTo` an "interim" branch. You could write custom automation to open a PR from the interim branch to the live branch or look at [GitOps Promoter](https://github.com/argoproj-labs/gitops-promoter) or [Kargo](https://github.com/akuity/kargo) which both have features around that.

imo the gold standard will be having first-class previews as PR comments so you don't have to merge and wait for the hydrator to run in order to get your first look at the rendered manifests.
